llama_index/llms/openai.py

```diff
@@ -242,25 +242,32 @@ def gen() -> ChatResponseGen:
                 content_delta = delta.content or ""
                 content += content_delta
 
-                function_call_delta = delta.function_call
-                if function_call_delta is not None:
-                    function_dict = function_call_delta.dict()
-
-                    if function_call is None:
-                        function_call = function_dict
-
-                        ## ensure we do not add a blank function call
-                        if function_call.get("function_name", "") is None:
-                            del function_call["function_name"]
+                tool_calls_delta = delta.tool_calls or None
+                if tool_calls_delta is not None:
+                    if tool_calls is None:
+                        tool_calls = [t.dict() for t in tool_calls_delta]
                     else:
-                        function_call["arguments"] = (
-                            function_call.get("arguments", "")
-                            + function_dict["arguments"]
-                        )
+                        for ix, t in enumerate(tool_calls):
```

Comment:
ooh ya this does get quite a bit more nuanced. Looks good at high-level to me though.

Reply:
yeah, i re-factored a bit and this is now accumulating the deltas thru working with `ChoiceDeltaToolCall` objects throughout. Its only at the very end do I convert this into a List[dict] as the downstream expects. Improved readability and speed.